src/wallet/rpcwallet.cpp

```diff
@@ -305,6 +305,289 @@ UniValue sendtoaddress(const UniValue& params, bool fHelp)
+
+    /// sapling addresses
+    std::set<libzcash::SaplingPaymentAddress> saplingAddresses;
+    pwalletMain->GetSaplingPaymentAddresses(saplingAddresses);
+
+    // legacy_random source
+    {
+        // Add legacy randomly generated address records to the result.
+        // This includes transparent addresses generated by the wallet via
+        // the keypool and Sprout addresses for which we have the
+        // spending key.
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "legacy_random");
+        bool hasData = false;
+
+        UniValue random_t(UniValue::VOBJ);
+
+        if (!t_generated_dests.empty()) {
+            UniValue random_t_addrs(UniValue::VARR);
+            for (const CTxDestination& dest : t_generated_dests) {
+                random_t_addrs.push_back(keyIO.EncodeDestination(dest));
+            }
+            random_t.pushKV("addresses", random_t_addrs);
+            hasData = true;
+        }
+
+        if (!t_change_dests.empty()) {
+            UniValue random_t_change_addrs(UniValue::VARR);
+            for (const CTxDestination& dest : t_change_dests) {
+                random_t_change_addrs.push_back(keyIO.EncodeDestination(dest));
+            }
+            random_t.pushKV("change_addresses", random_t_change_addrs);
+            hasData = true;
+        }
+
+        if (!t_generated_dests.empty() || !t_change_dests.empty()) {
+            entry.pushKV("transparent", random_t);
+        }
+
+        if (!sproutAddresses.empty()) {
+            UniValue random_sprout_addrs(UniValue::VARR);
+            for (const SproutPaymentAddress& addr : sproutAddresses) {
+                if (HaveSpendingKeyForPaymentAddress(pwalletMain)(addr)) {
+                    random_sprout_addrs.push_back(keyIO.EncodePaymentAddress(addr));
+                }
+            }
+
+            UniValue random_sprout(UniValue::VOBJ);
+            random_sprout.pushKV("addresses", random_sprout_addrs);
+
+            entry.pushKV("sprout", random_sprout);
+            hasData = true;
+        }
+
+        if (hasData) {
+            ret.push_back(entry);
+        }
+    }
+
+    /// imported source
+    {
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "imported");
+
+        bool hasData = false;
+        {
+            UniValue imported_sapling_addrs(UniValue::VARR);
+            for (const SaplingPaymentAddress& addr : saplingAddresses) {
+                if (GetSourceForPaymentAddress(pwalletMain)(addr) == Imported) {
+                    imported_sapling_addrs.push_back(keyIO.EncodePaymentAddress(addr));
+                }
+            }
+
+            if (!imported_sapling_addrs.empty()) {
+                UniValue imported_sapling_obj(UniValue::VOBJ);
+                imported_sapling_obj.pushKV("addresses", imported_sapling_addrs);
+                UniValue imported_sapling(UniValue::VARR);
+                imported_sapling.push_back(imported_sapling_obj);
```

Comment:
Hmm, if we're nesting here similarly to addrs with generated spending keys, then we should similarly group by viewing key.